### PR TITLE
8279283 - BufferedInputStream should override transferTo

### PR DESCRIPTION
**Pull request by mkarg**
Implementation of JDK-8279283

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8279283](https://bugs.openjdk.org/browse/JDK-8279283): BufferedInputStream should override transferTo

### Reviewers
 * @stsypanov (no known github.com user name / role) ⚠️ Review applies to [b4d5004d](https://git.openjdk.org/jdk/pull/6935/files/b4d5004dd3299d084769309b505b19ed7dd3182e)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/6935/head:pull/6935` \
`$ git checkout pull/6935`

Update a local copy of the PR: \
`$ git checkout pull/6935` \
`$ git pull https://git.openjdk.org/jdk pull/6935/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6935`

View PR using the GUI difftool: \
`$ git pr show -t 6935`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/6935.diff">https://git.openjdk.org/jdk/pull/6935.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
BIS is not specified to be thread safe but the existing read/write operations are. If transferTo is overridden then this is an area that will require close attention.

Have you surveyed the existing tests to see if transferTo is invoked on a BIS? New tests may be needed.

**Comment by mkarg** [comment]
> BIS is not specified to be thread safe but the existing read/write operations are. If transferTo is overridden then this is an area that will require close attention.

In analogy to the other read/write operations I now have synchronized transferTo in https://github.com/openjdk/jdk/pull/6935/commits/8dac240d2716a9d14c997ea62b7f96637acc8d66 to be on the safe side.

> Have you surveyed the existing tests to see if transferTo is invoked on a BIS? New tests may be needed.

Did not find an existing test for BIS.transferTo, so I will write a new test for this.

**Comment by mkarg** [comment]
> Have you surveyed the existing tests to see if transferTo is invoked on a BIS? New tests may be needed.

I have provided a test for BIS.transferTo in https://github.com/openjdk/jdk/pull/6935/commits/fbc5def7af6634b630c0bf423d8c7a2715b25cf4.

**Comment by stsypanov** [comment]
Maybe we need to include into this patch the benchmark referenced to in the body of the ticket?

**Comment by AlanBateman** [comment]
I think you'll need to look at the interaction with mark/reset. I don't think you can bypass the buffering when there is a mark set.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (dfuch) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:501]: Hi Mark, did you double check with all the existing subclasses of `BufferedInputStream` that the new implementation of `transferTo` will not cause trouble, if any of the subclasses is overriding `int read(byte[] b, int off, int len)`? If I'm not mistaken - with this new implementation of `transferTo` then `BufferedInputStream::read(byte[] b, int off, int len)` (a public method that can be overridden) will no longer be called by `transferTo`. This could be causing problems if `read` is overriden to do some special processing in the subclass.

**Comment by mkarg** [comment]
Good catches, I will look into your comments!

**Comment by mkarg** [comment]
Please keep open, still working on it.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by AlanBateman** [comment]
Can this PR be closed or returned to daft?

**Comment by mkarg** [comment]
I think we should turn it back to draft.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (stsypanov) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:495]: I think `buf` should not be accessed directly but via `getBufIfOpen()` because we need to throw IOE in case `this` is closed.

**Comment by stsypanov** [comment]
Will this be reopened somewhen? The proposed changes seems useful

**Comment by mkarg** [comment]
/open

I will continue work on this in the next weeks.

**Comment by mkarg** [comment]
> It's probably best to change to back to draft until there is an updated proposal.

I would love to, but how to do that? I mean, there is no button in Github that would allow *me* to do that.

**Comment by jaikiran** [comment]
> I mean, there is no button in Github that would allow me to do that.

On the right side of the page (at the top), search for "Still in progress?" "Convert to draft". That's an link (and I agree it's not easy to spot).

**Comment by mkarg** [comment]
> > I mean, there is no button in Github that would allow me to do that.
> 
> On the right side of the page (at the top), search for "Still in progress?" "Convert to draft". That's a link (and I agree it's not easy to spot)

Thanks, this worked! :-)

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (mkarg) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:501]: Thank you, good catch! I will check the subclasses contained in OpenJDK! 👍

**Comment by AlanBateman** [comment]
It looks like this patch is against a repository that hasn't been sync'ed up since late 2021. BIS has changed, esp. the locking, so this is why you get the merge-conflict label. Look at the implXXX methods to see how the existing methods do the locking.

I think I pointed out in one of the early rounds that you'll have to take the mark (if set) into account. It may be that you just call super.transferTo when markpos == -1.

The other issue that needs consideration is that the draining of the buffered bytes will leak the underlying input stream to the target output stream. It may be safer to also call super.transferTo when (count - pos) > 0.

File: test/lib/jdk/test/lib/hexdump/HexPrinter.java
Comment (dfuch) [inline test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: This is an indication that overriding `transferTo()` in `BufferedInputStream` has potential compatibility impacts on its subclasses. Therefore I would suggest adding a check in `BufferedInputStream::transferTo` to only override the behavior if not subclassed. Something like:

```
     if (this.getClass() != BufferedInputStream.class) {
         return super.transferTo(out);
     }
     ... otherwise proceed with the new implementation ...
```

Then you can revert this change here.

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: I do not see that it makes much sense to add such a safety means before I am finished with the inspection of the subclasses. As you can see in this example, it was pretty easy to fix it, and there are only few such subclasses at all. So instead of preparing against possibly non-existing easyt-to-fix problems, I prefer fixing the actual problems. They can only occur inside of the JDK itself, as relying on any particular implementation inside of the JDK by non-JDK classes simply would be a complete programming fault and *must* get fixed.

Comment (dfuch) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: It's a well known behavior that overriding the`read(...)` method is sufficient to implement subclass behavior. This will no longer be the case if `transferTo(...)` no longer calls `this.read(...)` as it used to. There are many undocumented behaviors that have been depended on over the year - and though I agree with the sentiment that it's not a good idea to depend on undocumented behavior, this is one that is long standing, and mostly legitimate, especially for those classes that were coded before `transferTo` was added to `InputStream`.

The issue here is that things would mostly work for those custom subclassed, up to the point where some unsuspecting code might call `transferTo`. This could lead to bugs that might become hard to diagnose, and of course would only be seen at runtime.
As a point of comparison, look at what [JEP 425](https://openjdk.org/jeps/425) did to change the locking in `BufferedInputStream` (see how the lock object is used/initialized).

How many classes of `BufferedInputStream` are there in the wild, outside of the JDK?
How many of them have custom behavior implemented in `read`, that would be broken if `transferTo` no longer called `this.read`?

If you're changing an observable behavior, you will need a CSR, you will need to evaluate the compatibility risk and justify it, you will need to write a Release Note to warn about the compatibility risk, so that custom implementations that have such a dependency can be changed to also override `transferTo`.

The main question is whether the advantage of changing the behavior outweigh the risk of regression it might cause in subclasses. Such might be the case - or not.
The middle ground - and safer path - is to override the behaviour only for the base class and those subclasses that do not depend on the default behavior as was done by JEP 425...

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: > It's a well known behavior that overriding the`read(...)` method is sufficient to implement subclass behavior. This will no longer be the case if `transferTo(...)` no longer calls `this.read(...)` as it used to. There are many undocumented behaviors that have been depended on over the year - and though I agree with the sentiment that it's not a good idea to depend on undocumented behavior, this is one that is long standing, and mostly legitimate, especially for those classes that were coded before `transferTo` was added to `InputStream`.
> 
> The issue here is that things would mostly work for those custom subclassed, up to the point where some unsuspecting code might call `transferTo`. This could lead to bugs that might become hard to diagnose, and of course would only be seen at runtime. As a point of comparison, look at what [JEP 425](https://openjdk.org/jeps/425) did to change the locking in `BufferedInputStream` (see how the lock object is used/initialized).
> 
> How many classes of `BufferedInputStream` are there in the wild, outside of the JDK? How many of them have custom behavior implemented in `read`, that would be broken if `transferTo` no longer called `this.read`?
> 
> If you're changing an observable behavior, you will need a CSR, you will need to evaluate the compatibility risk and justify it, you will need to write a Release Note to warn about the compatibility risk, so that custom implementations that have such a dependency can be changed to also override `transferTo`.
> 
> The main question is whether the advantage of changing the behavior outweigh the risk of regression it might cause in subclasses. Such might be the case - or not. The middle ground - and safer path - is to override the behaviour only for the base class and those subclasses that do not depend on the default behavior as was done by JEP 425...

I understand your claim but disagree with the conclusion. You could impose it to *any* behavior change of *any* class -- this is no special case here. I do not see that we need a CSR as I do not change the *documented* behavior (neither the JavaDoc nor the Spec). Also I need to ask back: Do you *really* assume there are such lots of people using `transferTo` with failing subclasses of BIS? I doubt that, actually. You try to fix things possible not broken at all. We should concentrate on facts, not on assumptions, IMHO.

Having said that, I leave it up to @AlanBateman to decide, as I have no strong feelings about the issue you raised.

Comment (dfuch) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: With the changes you proposed a CSR will definitely be needed.

Comment (AlanBateman) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: BIS dates from JDK 1.0 and it's not hard to find examples that extend it. The HexPrinter test reminds us that it has been possible for 25+ years to override the read methods and snoop on all bytes that are read. Adding an override of the transferTo method that bypasses the read methods might break these subclasses. So I think Daniel is right that we have to be cautious.

So I think the feedback that you have now is that bypassing the buffer is problematic when using mark, when there are buffered bytes, or when BIS has been extended. It may be that the starting point is something very conservative like this:

```
    @Override
    public long transferTo(OutputStream out) throws IOException {
        if (getClass() == BufferedInputStream.class
                && ((count - pos) <= 0) && (markpos == -1)) {
            return getInIfOpen().transferTo(out);
        } else {
            return super.transferTo(out);
        }
    }
```

There is also the issue of locking and async close to look into.

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: Thank you for your decision about this issue, so instead of fixing the existing bugs I will implement as you outlined.

BTW, locking and rebasing is on the way.

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: Rebased and fixed locking, using your proposed code now.

Async close leaves BIS in an invalid state. The JavaDocs say ` The behavior for the case where the inputand/or output stream is asynchronously closed, or the thread interrupted during the transfer, is highly input and output streamspecific, and therefore not specified.`. So there is seems to be no *essential* need to do something *particular* in that case (the caller cannot further use the BIS, just like after an IOE: `It is strongly recommended that both streams be promptly closed if an I/O error occurs.` -- which is what *I personally, as a user of BIS* would do after async close), but if you *want* that we reach a particular outcome, then just tell me what outcome you envision.

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: > With the changes you proposed a CSR will definitely be needed.

Just to learn about OpenJDK rules: Why? I always thought that CSR is only needed when breaking *documented* behavior. Do you have a link to the official rule that proofs your claim in *this* case (i. e. *non-documented* behavior)? Thanks! :-)

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (stsypanov) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:608]: Tiny thing: I think with 3 `&&`s we don't need parenthesis around `((count - pos) <= 0) && (markpos == -1))`. How about to leave leave one around `count - pos`?

File: test/lib/jdk/test/lib/hexdump/HexPrinter.java
Comment (dfuch) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: Hi Markus, It's a good idea to write a CSR if your change modifies an observable behavior that users of the public API (including custom subclasses) might be depending on. This is both to raise the awareness on the impact of the change, document the possible regression risks, and get feedback from the CSR committee who might suggest mitigating actions to reduce the risk (such as e.g: adding a property to make it possible to revert to the old behavior, write a release note, etc...). In the case at hand `BufferedInputStream` is a very old class that has been used and abused all over the board in the java ecosystem. So writing a CSR to document a change in behavior that some subclasses may be depending on would be the proper thing to do, when changing that behaviour. The fact that you have found such a subclass (with a legitimate use case) only strengthen the need for exercising such care.
In retrospect it would have been much better if `BufferedInputStream` had been declared final - but it's much too late for that now. Thanks!

Comment (AlanBateman) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: > @AlanBateman Async close leaves BIS in an invalid state. The JavaDocs say ` The behavior for the case where the input and/or output stream is asynchronously closed, or the thread interrupted during the transfer, is highly input and output stream specific, and therefore not specified.`. 

Yes, I believe I suggested that wording when the method was added. The java.io APIs mostly pre-date deep consideration of those topics.

I've looked at the update change to BIS in f2f9a904 and I think it's okay.

On the CSR topic and your messages with Daniel. On the CSR wiki pages then phrase to look for is "behavioral compatibility".  The "Kinds of Compatibility" has more details on the topic.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (mkarg) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:608]: I was using the exact code proposed by @AlanBateman and I don't dare to touch it. ;-)

Comment (mkarg) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:608]: @AlanBateman Please decide. :-)

File: test/lib/jdk/test/lib/hexdump/HexPrinter.java
Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: Thank you all for your kind help and feedback.

@AlanBateman So is now the time to switch this PR from Draft to Ready?

Comment (dfuch) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java:1194]: The source changes LGTM. I suppose you should now revert the changes above (in `HexPrinter.java`), otherwise the bytes will now be counted twice :-).
Which BTW shows that you would be depending on undocumented behavior whatever you do:
`transferTo` is not documented to call `this.read(...)`, but neither is it documented to **not call**  `this.read(...)`...  Hyrum's law wins here ;-)

Comment (mkarg) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java]: https://github.com/openjdk/jdk/pull/6935/commits/b4d5004dd3299d084769309b505b19ed7dd3182e reverts the change, as it is not needed (and wrong) with the check in place. I wonder if in future people will actually understand what actually is going on under the hood here.

Comment (AlanBateman) [reply test/lib/jdk/test/lib/hexdump/HexPrinter.java]: > @AlanBateman So is now the time to switch this PR from Draft to Ready?

Yes. From what I can tell, moving to Draft while you worked on it didn't pause the Skara bots, I'm not sure why that is.

**Comment by mkarg** [comment]
Kindly requesting approval of this PR. :-)

**Comment by mkarg** [comment]
I personally would rather go with `UncheckedIOException` instead of `IOException`, so the caller still can catch if he likes, but is not forced to catch-or-throw.

**Comment by mkarg** [comment]
Strange. Whenever I click on the "request review" icon, it removes the existing review requests?!

**Review by stsypanov — APPROVED** [review]
LGTM

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (jaikiran) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:2]: Hello Markus, the copyright year here should be `2022` instead of `2021`

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (jaikiran) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: Given that the implementation in this PR now calls the wrapped `InputStream`'s `transferTo` method, should we also include a test where the wrapped `InputStream` is a `FileInputStream` and the target `OutputStream` is a `FileOutputStream`, so that it exercises the `FileChannel#transferTo` code path and verifies the correct data is transferred?

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:2]: Actually I wrote this code in 2021.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: The correctness of the actually written data is already checked in `checkTransferredContents`. I do not see what you win by replacing `ByteArrayInputStream` by `FileInputStream`. What additional risks do you cover, and what is so special about `FileInputStream` compared to any other `InputStream`? `FileInputStream::transferTo` is already checked by its own unit test.

Comment (AlanBateman) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: > Given that the implementation in this PR now calls the wrapped `InputStream`'s `transferTo` method, should we also include a test where the wrapped `InputStream` is a `FileInputStream` and the target `OutputStream` is a `FileOutputStream`, so that it exercises the `FileChannel#transferTo` code path and verifies the correct data is transferred?

Right, the implementation change means that the transferTo implementation of the wrapped input stream is used, or the base InputStream.transferTo is used. We should expect there is good test coverage of these implementations already but checking into that might find opportunities to expand those tests (not this PR of course).

For the new test to exercise the BIS.transferTo then I think it will need to test with a mark set, test when there are buffered bytes (read before transferTo), or when BIS has been extended.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: Agreed that testing the non-empty-buffer (read-before-transferTo) and the mark-set cases (mark-before-transferTo), but still I do not see any need to take particularly `FileInputStream` into the boat; files only make the test run slower.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: I have just added testing `transferTo` with non-empty buffer and mark set as part of the already existing *randomized* test steps. I think that should be sufficient to detect problems if they really would exist, so there should not be a real need to run *explicit and separate* test cases for the combinations of "buffered with mark" / "buffered without mark" / "unbuffered with mark" / "unbuffered without mark" as that would not provide any improved coverage or detection rate IMHO.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (jaikiran) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:2]: > Actually I wrote this code in 2021. What are the official rules here, do I have to use the original date or the latest change date?

I remember reading it somewhere but I can't seem  to locate the text right now and it isn't mentioned in the OpenJDK developer's guide either https://openjdk.org/guide/.
From what I remember, the copyright year is of the form `year1, [year2]` where `year1` is the year the file was first added to the project (so in this case it would be `2022`) and `year2` is the year it was last updated. That's my understanding of the guidelines. Others can correct me if that isn't accurate.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (jaikiran) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: My intention of suggesting a test where it uses something like a FileInputStream was to trigger the code path that would invoke the `FileChannel#transferTo` when such an `InputStream` was wrapped in a `BufferedInputStream`. The `FileChannel#transferTo` in its javadoc states that it can throw exceptions that aren't `IOException` (for example `NonReadableChannelException` which is a `IllegalStateException`). So it was just an attempt to see if such a testing uncovers something that might make us consider catching those exceptions in this new implementation of `BufferedInputStream#transferTo` and rethrow it as an `IOException`.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:105]: The original implementation of `transferTo()` doesn't do that either, so if you think `transferTo()` must do that, that has to go into the *original* implementation, not only into the one of BIS in particular. To keep things simple, I'd propose that we do not mix ideas and you file a new issue / PR for that.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:2]: Correct, so 2021 is right here, as the second date (last change) is *optional*.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:2]: At least "added to the project" would be "merged", not "opened pr".

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (AlanBateman) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:170]: This just tests transferTo with a mark set, it doesn't appear to check that mark is respected. To test mark/reset then I think you'll see it mark with a largish readLimit, transferTo to drain the input to EOF, reset, and then read or transferTo again to check that the bytes from the mark are replayed.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:170]: I did not do that so far, as we agreed to check the *new* code path, which is only executed when noting was marked (so reset plays no role). Certainly I will be happy to add another test for the use case you describe (i. e. the *old* code path), even if we all know -thanks to our knowledge of the existence of the two code paths-, that it will surely pass even with my PR in place, unless there is a bug in the *old* code path... ;-)

Comment (dfuch) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:170]: Unless there's a bug and the new code path is triggered when there's a mark ;-) . Maybe we can convince ourselves that this can't happen now (by reading the code), but if someone modifies this code 2 years from now and introduces a bug that changes that it would be good to have a test to catch it.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:170]: It depends on the target of a test, but I am fine with adding such a test, as I alrready wrote.

Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java:170]: Since https://github.com/openjdk/jdk/pull/6935/commits/5d24268a807faa4535ea66f7e40ca6e4b983ce51 the test now proofs that mark/reset is respected by subsequent transferTo.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java]: Updated copyright to 2022 in https://github.com/openjdk/jdk/pull/6935/commits/a4b6f9130a045064370e49c0678429b26616fc3d as per your request.

**Comment by mkarg** [comment]
@AlanBateman I do not quite understand what would be wrong with the code below instead of falling back to the super implementation *in case of non-empty buffer*?
```java
private long implTransferTo(OutputStream out) throws IOException {
    if (getClass() == BufferedInputStream.class && markpos < 0) {
        int avail = count - pos;
        if (avail > 0) {
            out.write(getBufIfOpen(), pos, avail);
            pos = count;
        }
        return avail + getInIfOpen().transferTo(out);
    } else {
        return super.transferTo(out);
    }
}
```
Can you please explain why you proposed to use the super implementation for non-empty buffer? Just for my understanding. Thanks!

**Comment by AlanBateman** [comment]
> I do not quite understand what would be wrong with the code below instead of falling back to the super implementation _in case of non-empty buffer_?

I think you are asking if is safe to leak a reference to the internal buffer. If there is no mark then it might be okay because there is no replay for an evil output stream to attack. However, I think it would require wider review to be confident that there aren't other interesting ways to break it; hence the suggestion in one of the earlier comments to keep it simple and limit it when there is no subclassing, no mark, and no bytes buffered. This does not prevent widening the conditions in the future.

**Comment by mkarg** [comment]
> > I do not quite understand what would be wrong with the code below instead of falling back to the super implementation _in case of non-empty buffer_?
> 
> I think you are asking if is safe to leak a reference to the internal buffer. If there is no mark then it might be okay because there is no replay for an evil output stream to attack. However, I think it would require wider review to be confident that there aren't other interesting ways to break it; hence the suggestion in one of the earlier comments to keep it simple and limit it when there is no subclassing, no mark, and no bytes buffered. This does not prevent widening the conditions in the future.

Thank you. So I will file another PR for such a further discussion once we merged the current one.

**Comment by mkarg** [comment]
@AlanBateman Kindly requesting approval. :-)

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (bplb) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:40]: This import of `StandardOpenOption` is not needed due to line 55.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (bplb) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:89]: A number of code lines in the test are too long and make it even harder to read.

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (bplb) [inline test/jdk/java/io/BufferedInputStream/TransferTo.java:162]: transformTo -> transferTo

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java]: Done :-)

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java]: Done :-)

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java]: Done :-)

File: test/jdk/java/io/BufferedInputStream/TransferTo.java
Comment (mkarg) [reply test/jdk/java/io/BufferedInputStream/TransferTo.java]: Done :-)

**Comment by mkarg** [comment]
@bplb Fixed the issues you pointed out. Kindly requesting approval. :-)

**Comment by AlanBateman** [comment]
> @AlanBateman Kindly requesting approval. :-)

I don't have any more comments, the recent test cleanups are good. I assume @bplb will sponsor.

**Comment by mkarg** [comment]
> > @AlanBateman Kindly requesting approval. :-)
> 
> I don't have any more comments, the recent test cleanups are good. I assume @bplb will sponsor.

Thanks a lot for all your kind help with this PR! :-)

**Comment by mkarg** [comment]
@AlanBateman Kindly asking for `/approve`. Thanks. :-)

**Comment by bplb** [comment]
> @bplb Fixed the issues you pointed out.

@mkarg You are welcome.

**Review by bplb — APPROVED** [review]

**Comment by mkarg** [comment]
@bplb I would be very glad if you would `/sponsor` my PR. :-)

**Comment by bplb** [comment]
> @bplb I would be very glad if you would `/sponsor` my PR. :-)

Done.

**Comment by mkarg** [comment]
> I think you are asking if is safe to leak a reference to the internal buffer. If there is no mark then it might be okay because there is no replay for an evil output stream to attack. However, I think it would require wider review to be confident that there aren't other interesting ways to break it; hence the suggestion in one of the earlier comments to keep it simple and limit it when there is no subclassing, no mark, and no bytes buffered. This does not prevent widening the conditions in the future.

@AlanBateman I opened [another PR](https://github.com/openjdk/jdk/pull/10525) to continue this discussion.

**Pull request closed without merging**
